File: cosmoauthorization/src/java/it/csi/cosmo/cosmoauthorization/business/service/GruppiService.java
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package it.csi.cosmo.cosmoauthorization.business.service;

import java.util.List;
import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaGruppoRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaGruppoRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.GruppiResponse;
import it.csi.cosmo.cosmoauthorization.dto.rest.Gruppo;
import it.csi.cosmo.cosmoauthorization.dto.rest.RiferimentoGruppo;

/**
 *
 */

public interface GruppiService {

  /**
   * Restituisce il gruppo il cui id e' quello inserito nel database
   *
   * @param id del gruppo da cercare sul database
   * @return il gruppo del database
   */
  Gruppo getGruppo(Long id);


  /**
   * Restituisce il gruppo il cui codice e' quello inserito nel database
   *
   * @param codice del gruppo da cercare sul database
   * @return il gruppo del database
   */

  Gruppo getGruppoPerCodice(String codice);
  /**
   * Restituisce una lista di gruppo validi, il cui numero di elementi e' pari al valore indicato da
   * filter o, se filter non e' valorizzato, un numero di elementi pari al valore definito nella
   * tabella di configurazione
   *
   * @param filter indica i filtri per la ricerca dei gruppo
   * @return una lista di gruppo validi
   */
  GruppiResponse getGruppi(String filter);

  /**
   * Metodo che crea un gruppo sul database
   *
   * @param gruppo e' il gruppo che deve essere creato sul database
   * @return il gruppo creato sul database
   */
  Gruppo createGruppo(CreaGruppoRequest gruppo);

  /**
   * Metodo che aggiorna un gruppo sul database
   *
   * @param gruppo e' il gruppo che deve essere aggiornato sul database
   * @return il gruppo aggiornato sul database
   */
  Gruppo updateGruppo(Long id, AggiornaGruppoRequest gruppo);

  /**
   * @param id
   */
  void deleteGruppo(Long id);

  List<RiferimentoGruppo> getGruppiUtenteCorrente();
}

File: cosmowcl/src/app/shared/services/gruppi.service.ts
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { AggiornaGruppoRequest } from '../models/api/cosmoauthorization/aggiornaGruppoRequest';
import { CreaGruppoRequest } from '../models/api/cosmoauthorization/creaGruppoRequest';
import { GruppiResponse } from '../models/api/cosmoauthorization/gruppiResponse';
import { Gruppo } from '../models/api/cosmoauthorization/gruppo';
import { GruppoResponse } from '../models/api/cosmoauthorization/gruppoResponse';
import { RiferimentoGruppo } from '../models/api/cosmoauthorization/riferimentoGruppo';
import { ApiUrls } from '../utilities/apiurls';

@Injectable()
export class GruppiService {

  constructor(
    private http: HttpClient
  ) {
    // NOP
  }

  /** GET: Restituisce un gruppo. */
  getGruppo(id: number): Observable<GruppoResponse> {
    return this.http.get<GruppoResponse>(ApiUrls.GET_GRUPPO.replace('{id}', id.toString()));
  }

  /** GET: Restituisce una lista di gruppi. */
  getGruppi(filter: string): Observable<GruppiResponse> {
    const options = filter ?
      { params: new HttpParams().set('filter', filter) } : {};
    return this.http.get<GruppiResponse>(ApiUrls.GRUPPI, options);
  }

    /** GET: Restituisce una lista di gruppi. */
    getGruppiUtenteCorrente(): Observable<RiferimentoGruppo[]> {
      return this.http.get<RiferimentoGruppo[]>(ApiUrls.GRUPPI_UTENTE_CORRENTE);
    }

  /**
   * POST: Salva un gruppo.
   *
   * @param gruppo gruppo da salvare.
   * @returns Observable<Gruppo> il gruppo salvato.
   */

  creaGruppo(gruppo: CreaGruppoRequest): Observable<GruppoResponse> {
    return this.http.post<GruppoResponse>(ApiUrls.GRUPPI, gruppo);
  }

  /**
   * PUT: Aggiorna un gruppo.
   *
   * @param gruppo gruppo da aggiornare.
   * @returns Observable<Gruppo> il gruppo aggiornato.
   */

  aggiornaGruppo(id: number, gruppo: AggiornaGruppoRequest): Observable<GruppoResponse> {
    return this.http.put<GruppoResponse>(ApiUrls.GET_GRUPPO.replace('{id}', id.toString()), gruppo);
  }

  /** DELETE: Elimina un utente. */
  deleteGruppo(id: number): Observable<void> {
    return this.http.delete<void>(ApiUrls.GET_GRUPPO.replace('{id}', id.toString()));
  }
}
